test/core/tsi/s2a/s2a_handshaker_client_test.cc

```diff
@@ -0,0 +1,287 @@
+  tsi_handshaker_destroy(config->handshaker);
+  delete config;
+  config = nullptr;
+}
+
+static s2a_handshaker_client_config* s2a_handshaker_client_config_setup(
+    bool is_client) {
+  s2a_handshaker_client_config* config = new s2a_handshaker_client_config();
+  config->tsi_config = s2a_tsi_handshaker_config_setup(
+      is_client, kS2AHandshakerServiceUrlForTesting);
+  tsi_result result = s2a_handshaker_client_create(
+      reinterpret_cast<s2a_tsi_handshaker*>(config->tsi_config->handshaker),
+      config->tsi_config->channel,
+      /* interested_parties=*/nullptr, config->tsi_config->options,
+      grpc_slice_from_static_string(kS2AHandshakerClientTestTargetName),
+      /* grpc_cb=*/nullptr,
+      /* cb=*/nullptr,
+      /* user_data=*/nullptr, is_client, &(config->client));
+  GPR_ASSERT(result == TSI_OK);
+  return config;
+}
+
+static void s2a_handshaker_client_config_destroy(
+    s2a_handshaker_client_config* config) {
+  if (config == nullptr) {
+    return;
+  }
+  s2a_tsi_handshaker_config_destroy(config->tsi_config);
+  s2a_handshaker_client_destroy(config->client);
+  delete config;
+  config = nullptr;
+}
+
+static void s2a_handshaker_client_bad_options_test() {
+  s2a_tsi_handshaker_config* config =
+      s2a_tsi_handshaker_config_setup(/* is_client=*/true,
+                                      /* handshaker_service_url=*/nullptr);
+  s2a_handshaker_client* client = nullptr;
+  tsi_result result = s2a_handshaker_client_create(
+      reinterpret_cast<s2a_tsi_handshaker*>(config->handshaker),
+      config->channel,
+      /* interested_parties=*/nullptr, config->options,
+      grpc_slice_from_static_string(kS2AHandshakerClientTestTargetName),
+      /* grpc_cb=*/nullptr,
+      /* cb=*/nullptr,
+      /* user_data=*/nullptr,
+      /* is_client=*/true, &client);
+  GPR_ASSERT(result == TSI_INVALID_ARGUMENT);
+  GPR_ASSERT(client == nullptr);
+
+  // Cleanup.
+  s2a_tsi_handshaker_config_destroy(config);
+}
+
+static void s2a_handshaker_client_create_and_destroy_test() {
+  s2a_handshaker_client_config* client_config =
+      s2a_handshaker_client_config_setup(/* is_client=*/true);
+  s2a_handshaker_client_config* server_config =
+      s2a_handshaker_client_config_setup(/* is_client=*/false);
+
+  // Cleanup.
+  s2a_handshaker_client_config_destroy(client_config);
+  s2a_handshaker_client_config_destroy(server_config);
+}
+
+static void s2a_handshaker_client_client_start_test() {
+  s2a_handshaker_client_config* config =
+      s2a_handshaker_client_config_setup(/* is_client=*/true);
+  tsi_result result = config->client->client_start();
+  /** The |result| should be TSI_UNIMPLEMENTED because the |make_grpc_call| API
+   *  currently unimplemented. In order to check that |client_start| was
+   *  successful, we verify that the |send_buffer_| field of |config->client| is
+   *  correct. **/
+  // TODO(mattstev): change the check to |result == TSI_OK| once
+  // |make_grpc_call| has been implemented.
```

Comment:
Once `make_grpc_call` is implemented, do we still need to check the `send_buffer_` for correctness or there would be another way to do that?

Reply:
We should still check `send_buffer_` for correctness once `make_grpc_call` is implemented. (You can see the analogous tests in #23 if you're interested.)